File: android/app/src/main/java/ai/cbm/capture/domain/model/CaptureContract.kt
package ai.cbm.capture.domain.model

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/**
 * Wire types for `POST /cbm/capture`, mirroring `contract/capture-metadata.schema.json` and the
 * iOS `CaptureContract.swift` field for field.
 *
 * Inert data classes by design: the domain logic lives in [ai.cbm.capture.domain.imaging.ImageTransform]
 * and [ai.cbm.capture.domain.intrinsics.IntrinsicsGate], so a contract change is a diff in one file.
 */
object CaptureContract {
    const val SCHEMA_VERSION = "1.0.0"
}

@Serializable
enum class IntrinsicsSource {
    ARKIT, ARCORE, ANDROID_CAMERA2, EXIF, MANUAL_OVERRIDE;

    val displayName: String
        get() = when (this) {
            ARKIT -> "ARKit factory calibration"
            ARCORE -> "ARCore factory calibration"
            ANDROID_CAMERA2 -> "Camera2 factory calibration"
            EXIF -> "EXIF estimate"
            MANUAL_OVERRIDE -> "Manual override"
        }
}

@Serializable
enum class TrackingState { NORMAL, LIMITED, NOT_AVAILABLE }

@Serializable
data class CaptureMetadata(
    @SerialName("schema_version") val schemaVersion: String = CaptureContract.SCHEMA_VERSION,
    @SerialName("capture_id") val captureId: String,
    @SerialName("building_id") val buildingId: String,
    @SerialName("reporter_email") val reporterEmail: String? = null,
    val description: String? = null,
    /** RFC 3339, UTC, taken at shutter time - not at upload time. */
    @SerialName("captured_at") val capturedAt: String,
    val client: ClientInfo,
    val image: ImageDescriptor,
    val camera: CameraIntrinsics,
    val target: TargetDescriptor,
    val pose: PoseSample? = null
)

@Serializable
data class ClientInfo(
    val platform: String = "ANDROID",
    @SerialName("app_version") val appVersion: String,
    @SerialName("os_version") val osVersion: String,
    @SerialName("device_model") val deviceModel: String
)

@Serializable
data class ImageDescriptor(
    val width: Int,
    val height: Int,
    @SerialName("mime_type") val mimeType: String = "image/jpeg",
    val sha256: String,
    @SerialName("byte_length") val byteLength: Int,
    /** Quarter-turns clockwise already baked into the pixels. */
    @SerialName("orientation_applied") val orientationApplied: Int,
    @SerialName("source_width") val sourceWidth: Int? = null,
    @SerialName("source_height") val sourceHeight: Int? = null,
    val scale: Double? = null
)

@Serializable
data class CameraIntrinsics(
    val source: IntrinsicsSource,
    val trusted: Boolean,
    val fx: Double,
    val fy: Double,
    val cx: Double,
    val cy: Double,
    val skew: Double = 0.0,
    val width: Int,
    val height: Int,
    val lens: String? = null,
    val distortion: List<Double>? = null
)

@Serializable
data class TargetDescriptor(
    val pixel: PixelPoint,
    val source: String = "USER_TAP",
    val centrality: Double? = null
)

@Serializable
data class PixelPoint(val x: Double, val y: Double)

@Serializable
data class PoseSample(
    val source: IntrinsicsSource,
    val position: Vector3,
    val rotation: QuaternionValue,
    @SerialName("tracking_state") val trackingState: TrackingState,
    @SerialName("coordinate_space") val coordinateSpace: String = "AR_SESSION_WORLD"
)

@Serializable
data class Vector3(val x: Double, val y: Double, val z: Double)

@Serializable
data class QuaternionValue(val x: Double, val y: Double, val z: Double, val w: Double)

// MARK: - Responses

@Serializable
data class CaptureAcceptedResponse(
    val ok: Boolean = true,
    @SerialName("capture_id") val captureId: String? = null,
    @SerialName("request_id") val requestId: String? = null,
    val status: String? = null,
    val duplicate: Boolean = false
)

@Serializable
data class CaptureErrorResponse(
    val ok: Boolean = false,
    val error: String,
    val detail: String? = null
)

@Serializable
data class CaptureHealthResponse(
    val ok: Boolean,
    @SerialName("building_id") val buildingId: String? = null,
    @SerialName("schema_version") val schemaVersion: String? = null
)

File: ios/CBMCapture/Core/Models/CaptureContract.swift
import Foundation

/// Wire types for `POST /cbm/capture`, mirroring `contract/capture-metadata.schema.json`.
///
/// These are deliberately dumb `Codable` value types with no behaviour. Domain logic lives in
/// `ImageTransform` and `IntrinsicsGate`; keeping the wire shape inert means a contract change
/// is a diff in one file rather than a hunt through the feature layer.
enum CaptureContract {
    static let schemaVersion = "1.0.0"
}

// MARK: - Provenance

enum IntrinsicsSource: String, Codable, Sendable, CaseIterable {
    case arkit = "ARKIT"
    case arcore = "ARCORE"
    case androidCamera2 = "ANDROID_CAMERA2"
    case exif = "EXIF"
    case manualOverride = "MANUAL_OVERRIDE"

    /// Human-facing label for the capture screen badge.
    var displayName: String {
        switch self {
        case .arkit: "ARKit factory calibration"
        case .arcore: "ARCore factory calibration"
        case .androidCamera2: "Camera2 factory calibration"
        case .exif: "EXIF estimate"
        case .manualOverride: "Manual override"
        }
    }
}

enum TargetSource: String, Codable, Sendable {
    case userTap = "USER_TAP"
}

enum TrackingState: String, Codable, Sendable {
    case normal = "NORMAL"
    case limited = "LIMITED"
    case notAvailable = "NOT_AVAILABLE"
}

// MARK: - Metadata document

struct CaptureMetadata: Codable, Sendable, Equatable {
    var schemaVersion: String = CaptureContract.schemaVersion
    var captureID: UUID
    var buildingID: String
    var reporterEmail: String?
    var description: String?
    var capturedAt: Date
    var client: ClientInfo
    var image: ImageDescriptor
    var camera: CameraIntrinsics
    var target: TargetDescriptor
    var pose: PoseSample?

    enum CodingKeys: String, CodingKey {
        case schemaVersion = "schema_version"
        case captureID = "capture_id"
        case buildingID = "building_id"
        case reporterEmail = "reporter_email"
        case description
        case capturedAt = "captured_at"
        case client, image, camera, target, pose
    }
}

struct ClientInfo: Codable, Sendable, Equatable {
    var platform: String = "IOS"
    var appVersion: String
    var osVersion: String
    var deviceModel: String

    enum CodingKeys: String, CodingKey {
        case platform
        case appVersion = "app_version"
        case osVersion = "os_version"
        case deviceModel = "device_model"
    }
}

struct ImageDescriptor: Codable, Sendable, Equatable {
    var width: Int
    var height: Int
    var mimeType: String = "image/jpeg"
    var sha256: String
    var byteLength: Int
    /// Quarter-turns clockwise already baked into the pixel data.
    var orientationApplied: Int
    var sourceWidth: Int?
    var sourceHeight: Int?
    var scale: Double?

    enum CodingKeys: String, CodingKey {
        case width, height, sha256, scale
        case mimeType = "mime_type"
        case byteLength = "byte_length"
        case orientationApplied = "orientation_applied"
        case sourceWidth = "source_width"
        case sourceHeight = "source_height"
    }
}

struct CameraIntrinsics: Codable, Sendable, Equatable {
    var source: IntrinsicsSource
    var trusted: Bool
    var fx: Double
    var fy: Double
    var cx: Double
    var cy: Double
    var skew: Double = 0
    var width: Int
    var height: Int
    var lens: String?
    var distortion: [Double]?
}

struct TargetDescriptor: Codable, Sendable, Equatable {
    var pixel: PixelPoint
    var source: TargetSource = .userTap
    var centrality: Double?
}

struct PixelPoint: Codable, Sendable, Equatable {
    var x: Double
    var y: Double
}

struct PoseSample: Codable, Sendable, Equatable {
    var source: IntrinsicsSource
    var position: Vector3
    var rotation: QuaternionValue
    var trackingState: TrackingState
    var coordinateSpace: String = "AR_SESSION_WORLD"

    enum CodingKeys: String, CodingKey {
        case source, position, rotation
        case trackingState = "tracking_state"
        case coordinateSpace = "coordinate_space"
    }
}

struct Vector3: Codable, Sendable, Equatable {
    var x: Double, y: Double, z: Double
}

struct QuaternionValue: Codable, Sendable, Equatable {
    var x: Double, y: Double, z: Double, w: Double
}

// MARK: - Responses

struct CaptureAcceptedResponse: Codable, Sendable {
    var ok: Bool
    var captureID: UUID?
    var requestID: UUID?
    var status: String?
    var duplicate: Bool?

    enum CodingKeys: String, CodingKey {
        case ok, status, duplicate
        case captureID = "capture_id"
        case requestID = "request_id"
    }
}

struct CaptureErrorResponse: Codable, Sendable {
    var ok: Bool
    var error: String
    var detail: String?
}

struct CaptureHealthResponse: Codable, Sendable {
    var ok: Bool
    var buildingID: String?
    var schemaVersion: String?

    enum CodingKeys: String, CodingKey {
        case ok
        case buildingID = "building_id"
        case schemaVersion = "schema_version"
    }
}

// MARK: - JSON coders

extension JSONEncoder {
    /// The single encoder used for the metadata part. RFC 3339 with fractional seconds, in UTC.
    static var captureContract: JSONEncoder {
        let encoder = JSONEncoder()
        encoder.dateEncodingStrategy = .custom { date, encoder in
            var container = encoder.singleValueContainer()
            try container.encode(ISO8601DateFormatter.captureContract.string(from: date))
        }
        encoder.outputFormatting = [.sortedKeys]
        return encoder
    }
}

extension JSONDecoder {
    static var captureContract: JSONDecoder {
        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .custom { decoder in
            let raw = try decoder.singleValueContainer().decode(String.self)
            guard let date = ISO8601DateFormatter.captureContract.date(from: raw) else {
                throw DecodingError.dataCorruptedError(
                    in: try decoder.singleValueContainer(),
                    debugDescription: "Expected RFC 3339 timestamp, got \(raw)"
                )
            }
            return date
        }
        return decoder
    }
}

extension ISO8601DateFormatter {
    static let captureContract: ISO8601DateFormatter = {
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        formatter.timeZone = TimeZone(secondsFromGMT: 0)
        return formatter
    }()
}
